### PR TITLE
HOSTEDCP-1063: Disallow webhooks URLs targeting control plane services

### PR DESCRIPTION
**Pull request by muraee**
**What this PR does / why we need it**:
Disallow mutating/validating webhooks using URLs targeting any service in the hcp namespace, unless 'hypershift.io/allow-guest-webhooks' flag is present.

**Which issue(s) this PR fixes** *(optional, use `fixes #<issue_number>(, fixes #<issue_number>, ...)` format, where issue_number might be a GitHub issue, or a Jira story*:
Fixes #[HOSTEDCP-1063](https://issues.redhat.com/browse/HOSTEDCP-1063)

**Checklist**
- [x] Subject and description added to both, commit and PR.
- [x] Relevant issues have been referenced.
- [ ] This change includes docs. 
- [ ] This change includes unit tests.

### CODE REVIEW

File: control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go [hunk 1]
```diff
@@ -1655,6 +1658,68 @@ func (r *reconciler) ensureCloudResourcesDestroyed(ctx context.Context, hcp *hyp
...
+	if err := r.cpClient.List(ctx, cpServices, client.InNamespace(r.hcpNamespace)); err != nil {
+		return fmt.Errorf("failed to list control plane services")
+	}
+
+	// disallow all urls tageting services in the hcp namespace by default unless 'hypershift.io/allow-guest-webhooks' flag is present.
+	disallowedUrls := make([]string, 0)
+	for _, svc := range cpServices.Items {
+		if _, exist := svc.Labels["hypershift.io/allow-guest-webhooks"]; !exist {
+			disallowedUrls = append(disallowedUrls, fmt.Sprintf("https://%s:", svc.Name))
+			disallowedUrls = append(disallowedUrls, fmt.Sprintf("https://%s.%s.svc", svc.Name, svc.Namespace))
```
Comment (csrwng) [inline control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:1674]: also include `https://%s.%s.svc.cluster.local`

**Review by csrwng — COMMENTED** [review]
Just one comment, otherwise lgtm

File: control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go (hunk 1 above)
Comment (muraee) [reply control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:1674]: thanks, added

**Comment by muraee** [comment]
HCCO [logs ](https://gcsweb-ci.apps.ci.l2s4.p1.openshiftapps.com/gcs/origin-ci-test/pr-logs/pull/openshift_hypershift/2775/pull-ci-openshift-hypershift-main-e2e-aws/1677279130355240960/artifacts/e2e-aws/run-e2e/artifacts/TestCreateCluster_PreTeardownClusterDump/namespaces/e2e-clusters-z2gb5-example-vzmxk/core/pods/logs/hosted-cluster-config-operator-569c48469b-2n5kq-hosted-cluster-config-operator.log) show that `multus.openshift.io` webhook deployed by CNO is being deleted.
```
{"level":"info","ts":"2023-07-07T12:03:04Z","msg":"deleting validating webhook configuration with a disallowed url","controller":"resources","object":{"name":""},"namespace":"","name":"","reconcileID":"79add1a8-9eab-4671-bf3a-bac98c2cbfd7","webhook_name":"multus.openshift.io","disallowed_url":"https://multus-admission-controller.e2e-clusters-z2gb5-example-vzmxk.svc/validate"}
```

we need to add `hypershift.openshift.io/allow-guest-webhooks` label to the service here
https://github.com/openshift/cluster-network-operator/blob/master/bindata/network/multus-admission-controller/001-service.yaml

**Comment by muraee** [comment]
/hold
wait on https://github.com/openshift/cluster-network-operator/pull/1879

File: control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go [hunk 2]
```diff
@@ -1655,6 +1658,69 @@ func (r *reconciler) ensureCloudResourcesDestroyed(ctx context.Context, hcp *hyp
...
+				errs = append(errs, r.client.Delete(ctx, &configuration))
+				break
+			}
+		}
+	}
+
+	mutatingWebhookConfigurations := &admissionregistrationv1.MutatingWebhookConfigurationList{}
+	if err := r.client.List(ctx, mutatingWebhookConfigurations); err != nil {
+		errs = append(errs, err)
+		errors.NewAggregate(errs)
```
Comment (enxebre) [inline control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:1698]: what is this NewAggregate doing here?

File: control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go [hunk 3]
```diff
@@ -1655,6 +1658,69 @@ func (r *reconciler) ensureCloudResourcesDestroyed(ctx context.Context, hcp *hyp
...
+		return err
+	}
+
+	errs := make([]error, 0)
+	for _, configuration := range validatingWebhookConfigurations.Items {
+		for _, webhook := range configuration.Webhooks {
+			if webhook.ClientConfig.URL != nil && !isAllowedWebhookUrl(disallowedUrls, *webhook.ClientConfig.URL) {
+				log.Info("deleting validating webhook configuration with a disallowed url", "webhook_name", configuration.Name, "disallowed_url", *webhook.ClientConfig.URL)
+				errs = append(errs, r.client.Delete(ctx, &configuration))
+				break
```
Comment (enxebre) [inline control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:1690]: why break here?

**Comment by enxebre** [comment]
lgtm overall
can we include an e2e that tries to create a conflicting webhook and we see how we revert it?

File: control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go (hunk 3 above)
Comment (muraee) [reply control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:1690]: you can define multiple webhooks in the same Resource (configuration), so if we find a violating webhook we delete the whole resource and break, no need to check the rest of the webhooks.

File: control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go (hunk 2 above)
Comment (muraee) [reply control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:1698]: to include any errors if any from deleting validatingWebhookConfigurations previously https://github.com/openshift/hypershift/blob/06d38188866c703ae05ba0e35070993c92c6bbdb/control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go#L1689

Comment (csrwng) [reply control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:1698]: but it's not being assigned to anything

Comment (muraee) [reply control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:1698]: ahhh, forgot the return :)

Comment (muraee) [reply control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go]: thanks, fixed

File: control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go [hunk 4]
```diff
@@ -1655,6 +1658,69 @@ func (r *reconciler) ensureCloudResourcesDestroyed(ctx context.Context, hcp *hyp
...
+			}
+		}
+	}
+
+	mutatingWebhookConfigurations := &admissionregistrationv1.MutatingWebhookConfigurationList{}
+	if err := r.client.List(ctx, mutatingWebhookConfigurations); err != nil {
```
Comment (bryan-cox) [inline control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:1696]: This reinitializes `err` so you would lose any errors captured on L1689 right?

Comment (muraee) [reply control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:1696]: why would I lose any errors?  `err` is not defined anywhere before this line.

Comment (bryan-cox) [reply control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:1696]: Yeah you're right. I misread L1684 as `err` 🤦🏻

**Review by bryan-cox — APPROVED** [review]
/lgtm

**Comment by muraee** [comment]
/unhold
https://github.com/openshift/cluster-network-operator/pull/1879 has been merged

File: control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go [hunk 5]
```diff
@@ -1655,6 +1658,69 @@ func (r *reconciler) ensureCloudResourcesDestroyed(ctx context.Context, hcp *hyp
 	return remaining, errors.NewAggregate(errs)
 }
 
+func (r *reconciler) ensureGuestAdmissionWebhooksAreValid(ctx context.Context) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	cpServices := &corev1.ServiceList{}
+	if err := r.cpClient.List(ctx, cpServices, client.InNamespace(r.hcpNamespace)); err != nil {
+		return fmt.Errorf("failed to list control plane services")
```
Comment (enxebre) [inline control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:1666]: forgot to pass the err here? `return fmt.Errorf("failed to list control plane services: %w", err)

File: control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go [hunk 6]
```diff
@@ -1655,6 +1658,69 @@ func (r *reconciler) ensureCloudResourcesDestroyed(ctx context.Context, hcp *hyp
...
+	}
+
+	// disallow all urls tageting services in the hcp namespace by default unless 'hypershift.openshift.io/allow-guest-webhooks' label is present.
+	disallowedUrls := make([]string, 0)
+	for _, svc := range cpServices.Items {
+		if _, exist := svc.Labels["hypershift.openshift.io/allow-guest-webhooks"]; !exist {
```
Comment (enxebre) [inline control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:1672]: should we put this here https://github.com/openshift/hypershift/blob/main/api/v1beta1/hostedcluster_types.go#L17?

Comment (enxebre) [inline control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:1672]: nit: I'd rather return early if exists than putting the whole logic into the if block

File: control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go [hunk 7]
```diff
@@ -1655,6 +1658,69 @@ func (r *reconciler) ensureCloudResourcesDestroyed(ctx context.Context, hcp *hyp
...
+			disallowedUrls = append(disallowedUrls, fmt.Sprintf("https://%s:", svc.Name))
+			disallowedUrls = append(disallowedUrls, fmt.Sprintf("https://%s.%s.svc", svc.Name, svc.Namespace))
+			disallowedUrls = append(disallowedUrls, fmt.Sprintf("https://%s.%s.svc.cluster.local", svc.Name, svc.Namespace))
+		}
+	}
+
+	validatingWebhookConfigurations := &admissionregistrationv1.ValidatingWebhookConfigurationList{}
+	if err := r.client.List(ctx, validatingWebhookConfigurations); err != nil {
+		return err
```
Comment (enxebre) [inline control-plane-operator/hostedclusterconfigoperator/controllers/resources/resources.go:1681]: nit: `return fmt.Errorf("failed to list validatingWebhookConfigurations: %w", err)`

File: test/e2e/util/util.go [hunk 8]
```diff
@@ -1252,3 +1255,54 @@ func EnsurePodsWithEmptyDirPVsHaveSafeToEvictAnnotations(t *testing.T, ctx conte
...
+				},
+			}},
+			SideEffects: &sideEffectsNone,
+		}}
+
+		if err := guestClient.Create(ctx, guestWebhookConf); err != nil {
+			t.Errorf("failed to create webhook: %v", err)
+		}
+
+		err := wait.PollImmediateWithContext(ctx, 5*time.Second, 5*time.Minute, func(ctx context.Context) (done bool, err error) {
```
Comment (enxebre) [inline test/e2e/util/util.go:1293]: wouldn't it be a bug if this takes for than a few secs? why 5min?

**Comment by enxebre** [comment]
Dropped some minor comments 
/lgtm

File: test/e2e/util/util.go (hunk 8 above)
Comment (muraee) [reply test/e2e/util/util.go]: right, I just didn't want to have flakes in CI because of this.
I changed it to 1min

**Pull request merged**